cpp/src/arrow/csv/writer.cc

```diff
@@ -0,0 +1,398 @@
+               std::vector<std::unique_ptr<ColumnPopulator>> populators, MemoryPool* pool)
+      : schema_(std::move(schema)),
+        column_populators_(std::move(populators)),
+        row_positions_(1024, nullptr, arrow::stl::allocator<char*>(pool)),
+        pool_(pool) {}
+
+  const std::shared_ptr<Schema> schema_;
+
+  Status PrepareForContentsWrite(const WriteOptions& options, io::OutputStream* out) {
+    if (data_buffer_ == nullptr) {
+      ASSIGN_OR_RAISE(
+          data_buffer_,
+          AllocateResizableBuffer(
+              options.batch_size * schema_->num_fields() * kColumnSizeGuess, pool_));
+    }
+    if (options.include_header) {
+      RETURN_NOT_OK(WriteHeader(out));
+    }
+    return Status::OK();
+  }
+
+  int64_t CalculateHeaderSize() const {
+    int64_t header_length = 0;
+    for (int col = 0; col < schema_->num_fields(); col++) {
+      const std::string& col_name = schema_->field(col)->name();
+      header_length += col_name.size();
+      header_length += CountEscapes(col_name);
+    }
+    return header_length + (3 * schema_->num_fields());
+  }
+
+  Status WriteHeader(io::OutputStream* out) {
+    RETURN_NOT_OK(data_buffer_->Resize(CalculateHeaderSize(), /*shrink_to_fit=*/false));
+    char* next = reinterpret_cast<char*>(data_buffer_->mutable_data());
+    for (int col = 0; col < schema_->num_fields(); col++) {
+      *next++ = '"';
+      next = Escape(schema_->field(col)->name(), next);
+      *next++ = '"';
+      *next++ = ',';
+    }
+    next--;
+    *next = '\n';
+    return out->Write(data_buffer_);
+  }
+
+  Status TranslateMininalBatch(const RecordBatch& batch) {
+    if (batch.num_rows() == 0) {
+      return Status::OK();
+    }
+    std::vector<int32_t, arrow::stl::allocator<int32_t>> offsets(
+        batch.num_rows(), 0, arrow::stl::allocator<int32_t>(pool_));
+
+    // Calculate relative offsets for each row (excluding delimiters)
+    for (size_t col = 0; col < column_populators_.size(); col++) {
+      RETURN_NOT_OK(
+          column_populators_[col]->UpdateRowLengths(*batch.column(col), offsets.data()));
+    }
+    // Calculate cumulalative offsets for each row (including delimiters).
+    offsets[0] += batch.num_columns();
+    for (int64_t row = 1; row < batch.num_rows(); row++) {
+      offsets[row] += offsets[row - 1] + /*delimiter lengths*/ batch.num_columns();
+    }
+    // Resize the target buffer to required size. We assume batch to batch sizes
+    // should be pretty close so don't shrink the buffer to avoid allocation churn.
+    RETURN_NOT_OK(data_buffer_->Resize(offsets.back(), /*shrink_to_fit=*/false));
+
+    // Calculate pointers to the start of each row.
+    row_positions_.resize(batch.num_rows());
+    row_positions_[0] = reinterpret_cast<char*>(data_buffer_->mutable_data());
+    for (size_t row = 1; row < row_positions_.size(); row++) {
+      row_positions_[row] =
+          reinterpret_cast<char*>(data_buffer_->mutable_data()) + offsets[row - 1];
```

Comment:
I'm curious: is it useful to compute a vector of `char*` positions rather than reuse the offsets array?